lua/neoclip/db.lua

```diff
@@ -0,0 +1,70 @@
+local M = {}
+
+local has_sqlite, sqlite = pcall(require, "sqlite")
+if not has_sqlite then
+    print "Couldn't find sqlite.lua. Cannot use persistent history"
+    return nil
+end
+
+local function get_db(db_path)
+    local db = sqlite.new(db_path)
+    db:open()
+    db:create('neoclip', {
+        id = true,
+        ensure = true,
+        regtype = 'text',
+        contents = 'text',
+        filetype = 'text',
+    })
+    return db
+end
+
+local function copy(t)
+    local new = {}
+    for k, v in pairs(t) do
+        if type(v) == 'table' then
+            new[k] = copy(v)
+        else
+            new[k] = v
+        end
+    end
+    return new
+end
+
+local function map(t, f)
+    local new_t = copy(t)
+    for k, v in pairs(new_t) do
+        new_t[k] = f(v)
+    end
+    return new_t
+end
+
+local function to_sqlite(storage)
+    local tbl = map(storage, function(v)
+        v.contents = table.concat(v.contents, '\n')
+        return v
+    end)
+    return tbl
+end
+
+local function from_sqlite(tbl)
+    local storage = map(tbl, function(v)
+        v.contents = vim.fn.split(v.contents, '\n')
+        return v
+    end)
+    return storage
+end
+
+M.get = function(db_path)
+    local db = get_db(db_path)
+    return from_sqlite(db:select('neoclip'))
+end
+
+M.update = function(opts)
+    local db = get_db(opts.db_path)
+    db:delete('neoclip')
+    -- TODO can arrays be used in sqlite columns? If so, then to_sqlite could be removed
+    db:insert('neoclip', to_sqlite(opts.storage))
```

Comment:
@tami5 when I run this line I get the following error:
```
E5108: Error executing lua ...im/site/pack/packer/start/sqlite.lua/lua/sqlite/stmt.lua:35: sqlite.lua: sql statement parse, , stmt: `ins
ert into neoclip (contents, filetype, regtype) values(    print(vim.inspect(to_sqlite(opts.storage))), :filetype, :regtype)`, err: `(`ne
ar "(": syntax error`)`
```
any idea what I'm doing wrong? The table is created as above with
```lua
    db:create('neoclip', {
        id = true,
        ensure = true,
        regtype = 'text',
        contents = 'text',
        filetype = 'text',
    })
```

and if I print `to_sqlite(opts.storage)` before the insertion it looks for example as:
```
{ {
    contents = "    print(vim.inspect(to_sqlite(opts.storage)))",
    filetype = "lua",
    regtype = "l"
  }, {
    contents = "    db_path = 'test.sql',",
    filetype = "lua",
    id = 1,
    regtype = "l"
  } }
```

Reply:
`    contents = "    print(vim.inspect(to_sqlite(opts.storage)))",`

I believe this is interpreted as sqlite function not a text value. hence added as is `values( print(vim.inspect(to_sqlite(opts.storage))), ..`

Reply:
The exact `contents` string is unimportant in this case, maybe confusing but I just copied some lines (using neoclip) which I had in the source file, it could have just as well looked like:
```
{ {
    contents = "some string",
    filetype = "lua",
    regtype = "l"
  }, {
    contents = "some other string",
    filetype = "lua",
    id = 1,
    regtype = "l"
  } }
```
Actually looking at it now, I'm not sure why the `id` key is there...

So what is the bug exactly? In sqlite.lua?

Reply:
hmmm this weird. update the last edit you made and I'll give it a try

Reply:
In the latest commit I changed to use `"luatables"` and a default `db_path`, however, I'm still having the same issue. If you want to try it you need to set `enable_persistant_history` to true in the neoclip setup function (or by
```
:lua require('neoclip.settings').get()['enable_persistant_history'] = true
```
)
copy some stuff and then do
```
:lua require('neoclip.storage').on_exit()
```
to manually trigger what happens on `VimLeavePre`.